# Multidimensionality of Aging - a tutorial on the Abdomen dimension
The project consist on predicting age using different types of medical images from UKB, and deep learning.
This pipeline generates the images-based predictors, then merges them and ensemble them with the predictors build on images. Finally, it performs postprocessing steps.


# Project architecture
There are six directories.

scripts: where the python scripts can be found. This is the core of the pipeline.

data: where the results used and generated by the pipeline are saved.

An exact list of the important file can be seen [here](##Structure:)
Before starting

The following are the steps that should be taken to run the pipeline: 01A, 01B, 01C, 01D, 02, 03A, 03B, 03C, 04A, 04B05C, 04C, 05A, 05B, 04B05C, 06A, 06B, 06C, 07A, 07B, 08, 09A, 09B, 09C, 09D, 09E, 09G
The function of these different steps is described below:

 MI01A_Preprocessing_main.sh
 Preprocesses the main dataframe.
 
 MI01B_Preprocessing_imagesIDs.sh
 Splits the different images datasets into folds for the future cross validation.

 MI01C_Preprocessing_folds_parallel.sh
 Splits the data into training, validation and testing sets for all CV folds.

 MI01D_Preprocessing_survival.sh
 Preprocesses the main dataframe for survival purposes.

 MI02_Training_parallel.sh
 Trains the models. This step should be repeated until all models have converged.

 MI03A_Predictions_generate_parallel.sh
 Generates the predictions from all models.

 MI03B_Predictions_concatenate_parallel.sh
 Concatenates the predictions from the different cross-validation folds.

 MI03C_Predictions_merge_parallel.sh
 Merges the predictions from all models into a unified dataframe.

 MI03D_Predictions_eids_parallel.sh
 Computes the average age prediction across samples from different instances for every participant.

 MI04A_Performances_generate_parallel.sh
 Computes the performances for each model.

 MI04B05C_Performances_merge_parallel.sh False
 Merges the performances of the different models into a unified dataframe.

 MI04C_Performances_tuning_parallel.sh
 For each model, selects the best hyperparameter combination.

 MI05A_Ensembles_predictions_generate_and_merge_parallel.sh
 Hierarchically builds ensemble models.

 MI05B_Ensembles_performances_generate_parallel.sh
 Computes the performances for the ensemble models.

 MI04B05C_Performances_merge_parallel.sh True
 Adds the ensemble models to the unified dataframe.

 MI06A_Residuals_generate_parallel.sh
 Computes accelerated aging phenotypes (Residuals, corrected for residuals bias with respect to age).

 MI06B_Residuals_correlations_parallel.sh
 Computes the phenotypic correlation between aging dimensions.

 MI06C_Performances_survival_parallel.sh
 Computes the performances in terms of survival prediction using biological age phenotypes as survival predictors.

 MI07A_Select_best_parallel.sh
 For each aging main dimension and selected subdimensions, select the best performing model.

 MI07B_Select_correlationsNAs_parallel.sh
 Build a summary correlation matrix: when a correlation cannot be computed in terms of samples ("instances") because the intersection has a small sample size, fill the NA with the correlation computed at the participant's level ("eids").

 MI08_Attentionmaps_parallel.sh
 Computes the attention maps (saliency maps and Grad_RAM maps) for all images.

 MI09A_GWAS_preprocessing_parallel.sh
 Preprocesses the data for the GWASs.

 MI09B_GWAS_remove_parallel.sh
 Runs a first pass of GWASs using BOLT to automatically generate the list of samples to remove for the actual GWASs.

 MI09C_GWAS_bolt_parallel.sh
 Performs the GWAS using BOLT to identify the SNPs, genes et heritability for each accelerated aging dimension.

 MI09D_GWAS_correlations_parallel.sh
 Computes the genetic correlations using BOLT.

 MI09E_GWAS_postprocessing_parallel.sh
 Postprocesses the GWAS results and stores the results in summary files.

 MI09F Annotation of the GWAS results
 /!\ Corresponds to a step in the pipeline that should be performed on local machine, since it must be complemented with researches on the internet for different steps. Refer to the python script MI09F_GWAS_annotate.py /!\ Annotates the hits from the GWAS: names of the genes and gene types.

 MI09G_GWAS_plots_parallel.sh
 Generates Manhattan and QQ plots to summarize the GWASs results.



## Structure :
 ğŸ“¦Age_Sex_and_Medical_Images
 â”£ ğŸ“‚.devcontainer
 â”ƒ â”£ ğŸ“œDockerfile
 â”ƒ â”— ğŸ“œdevcontainer.json
 â”£ ğŸ“‚data
 â”ƒ â”£ ğŸ“‚Abdomen
 â”ƒ â”ƒ â”£ ğŸ“‚Liver
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“‚Contrast
 â”ƒ â”ƒ â”ƒ â”ƒ â”£ ğŸ“œ1006879_2.jpg
 â”ƒ â”ƒ â”ƒ â”ƒ â”£ ğŸ“œ1008016_2.jpg
 â”ƒ â”ƒ â”ƒ â”ƒ â”— ğŸ“œ*.jpg
 â”ƒ â”ƒ â”ƒ â”— ğŸ“‚Raw
 â”ƒ â”ƒ â”ƒ   â”£ ğŸ“œ1006879_2.jpg
 â”ƒ â”ƒ â”ƒ   â”£ ğŸ“œ1008016_2.jpg
 â”ƒ â”ƒ â”ƒ   â”— ğŸ“œ*.jpg
 â”ƒ â”ƒ â”— ğŸ“‚Pancreas
 â”ƒ â”ƒ   â”£ ğŸ“‚Contrast
 â”ƒ â”ƒ   â”ƒ â”£ ğŸ“œ1013920_2.jpg
 â”ƒ â”ƒ   â”ƒ â”£ ğŸ“œ1023499_2.jpg
 â”ƒ â”ƒ   â”ƒ â”£ ğŸ“œ*.jpg
 â”ƒ â”ƒ   â”— ğŸ“‚Raw
 â”ƒ â”ƒ     â”£ ğŸ“œ1013920_2.jpg
 â”ƒ â”ƒ     â”£ ğŸ“œ1023499_2.jpg
 â”ƒ â”ƒ     â”— ğŸ“œ*.jpg
 â”ƒ â”£ ğŸ“‚FoldsAugmented
 â”ƒ â”ƒ â”£ ğŸ“œfake_data-features_Heart_20208_Augmented_Age_test_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œfake_data-features_Heart_20208_Augmented_Age_train_{0... 9}.csv
 â”ƒ â”ƒ â”— ğŸ“œfake_data-features_Heart_20208_Augmented_Age_val_{0... 9}.csv
 â”ƒ â”£ ğŸ“‚MI01A_Preprocessing_main
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_eids.csv
 â”ƒ â”ƒ â”— ğŸ“œdata-features_instances.csv
 â”ƒ â”£ ğŸ“‚MI01B_Preprocessing_imagesIDs
 â”ƒ â”ƒ â”£ ğŸ“œinstances23_eids_0.csv
 â”ƒ â”ƒ â”£ ğŸ“œinstances23_eids_1.csv
 â”ƒ â”ƒ â”£ ğŸ“œinstances23_eids_2.csv
 â”ƒ â”ƒ â”£ ğŸ“œinstances23_eids_3.csv
 â”ƒ â”ƒ â”£ ğŸ“œinstances23_eids_4.csv
 â”ƒ â”ƒ â”£ ğŸ“œinstances23_eids_5.csv
 â”ƒ â”ƒ â”£ ğŸ“œinstances23_eids_6.csv
 â”ƒ â”ƒ â”£ ğŸ“œinstances23_eids_7.csv
 â”ƒ â”ƒ â”£ ğŸ“œinstances23_eids_8.csv
 â”ƒ â”ƒ â”— ğŸ“œinstances23_eids_9.csv
 â”ƒ â”£ ğŸ“‚MI01C_Preprocessing_folds
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Liver_Contrast_Age_test_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Liver_Contrast_Age_train_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Liver_Contrast_Age_val_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Liver_Raw_Age_test_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Liver_Raw_Age_train_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Liver_Raw_Age_val_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Pancreas_Contrast_Age_test_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Pancreas_Contrast_Age_train_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Pancreas_Contrast_Age_val_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Pancreas_Raw_Age_test_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œdata-features_Abdomen_Pancreas_Raw_Age_train_{0... 9}.csv
 â”ƒ â”ƒ â”— ğŸ“œdata-features_Abdomen_Pancreas_Raw_Age_val_{0... 9}.csv
 â”ƒ â”£ ğŸ“‚MI02_Training
 â”ƒ â”ƒ â”£ ğŸ“œtrained_model-weights_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_{0... 9}.h5
 â”ƒ â”£ ğŸ“‚MI03A_Predictions_generate
 â”ƒ â”ƒ â”£ ğŸ“œshort_Predictions_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_test_{0... 9}.csv
 â”ƒ â”ƒ â”£ ğŸ“œshort_Predictions_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_train_{0... 9}.csv
 â”ƒ â”ƒ â”— ğŸ“œshort_Predictions_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_val_{0... 9}.csv
 â”ƒ â”£ ğŸ“‚MI03B_Predictions_concatenate
 â”ƒ â”ƒ â”£ ğŸ“œPredictions_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPredictions_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_train.csv
 â”ƒ â”ƒ â”— ğŸ“œPredictions_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_val.csv
 â”ƒ â”£ ğŸ“‚MI03C_Predictions_merge
 â”ƒ â”ƒ â”£ ğŸ“œPREDICTIONS_withoutEnsembles_instances_Age_test.csv
 â”ƒ â”ƒ â”— ğŸ“œPREDICTIONS_withoutEnsembles_instances_Age_val.csv
 â”ƒ â”£ ğŸ“‚MI03D_Predictions_eids
 â”ƒ â”ƒ â”£ ğŸ“‚Predictions_eids_concatenate
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œPredictions_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_test.csv
 â”ƒ â”ƒ â”ƒ â”— ğŸ“œPredictions_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPREDICTIONS_withoutEnsembles_eids_Age_test.csv
 â”ƒ â”ƒ â”— ğŸ“œPREDICTIONS_withoutEnsembles_eids_Age_val.csv
 â”ƒ â”£ ğŸ“‚MI04A_Performances_generate
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_test_sd.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_test_str.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_val_sd.csv
 â”ƒ â”ƒ â”— ğŸ“œPerformances_instances_Age_Abdomen_Pancreas_Contrast_InceptionResNetV2_1_1024_Adam_0.0001_0.1_0.5_1.0_val_str.csv
 â”ƒ â”£ ğŸ“‚MI04B_Performances_merge
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_withoutEnsembles_alphabetical_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_withoutEnsembles_alphabetical_instances_Age_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_withoutEnsembles_ranked_instances_Age_test.csv
 â”ƒ â”ƒ â”— ğŸ“œPERFORMANCES_withoutEnsembles_ranked_instances_Age_val.csv
 â”ƒ â”£ ğŸ“‚MI04C_Performances_tuning
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_tuned_alphabetical_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_tuned_alphabetical_instances_Age_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_tuned_ranked_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_tuned_ranked_instances_Age_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPREDICTIONS_tuned_instances_Age_test.csv
 â”ƒ â”ƒ â”— ğŸ“œPREDICTIONS_tuned_instances_Age_val.csv
 â”ƒ â”£ ğŸ“‚MI05A_Ensembles_predictions
 â”ƒ â”ƒ â”£ ğŸ“œPREDICTIONS_withEnsembles_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPREDICTIONS_withEnsembles_instances_Age_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPredictions_instances_Age_*_*_*_*_*_*_*_*_*_*_*_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPredictions_instances_Age_*_*_*_*_*_*_*_*_*_*_*_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPredictions_instances_Age_*instances01_*_*_*_*_*_*_*_*_*_*_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPredictions_instances_Age_*instances01_*_*_*_*_*_*_*_*_*_*_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPredictions_instances_Age_*instances1.5x_*_*_*_*_*_*_*_*_*_*_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPredictions_instances_Age_*instances1.5x_*_*_*_*_*_*_*_*_*_*_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPredictions_instances_Age_*instances23_*_*_*_*_*_*_*_*_*_*_test.csv
 â”ƒ â”ƒ â”— ğŸ“œPredictions_instances_Age_*instances23_*_*_*_*_*_*_*_*_*_*_val.csv
 â”ƒ â”£ ğŸ“‚MI05B_Performances_generate
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*_*_*_*_*_*_*_*_*_*_*_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*_*_*_*_*_*_*_*_*_*_*_test_sd.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*_*_*_*_*_*_*_*_*_*_*_test_str.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*_*_*_*_*_*_*_*_*_*_*_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*_*_*_*_*_*_*_*_*_*_*_val_sd.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*_*_*_*_*_*_*_*_*_*_*_val_str.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*instances23_*_*_*_*_*_*_*_*_*_*_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*instances23_*_*_*_*_*_*_*_*_*_*_test_sd.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*instances23_*_*_*_*_*_*_*_*_*_*_test_str.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*instances23_*_*_*_*_*_*_*_*_*_*_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPerformances_instances_Age_*instances23_*_*_*_*_*_*_*_*_*_*_val_sd.csv
 â”ƒ â”ƒ â”— ğŸ“œPerformances_instances_Age_*instances23_*_*_*_*_*_*_*_*_*_*_val_str.csv
 â”ƒ â”£ ğŸ“‚MI05C_Performances_merge
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_withEnsembles_alphabetical_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_withEnsembles_alphabetical_instances_Age_val.csv
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_withEnsembles_ranked_instances_Age_test.csv
 â”ƒ â”ƒ â”— ğŸ“œPERFORMANCES_withEnsembles_ranked_instances_Age_val.csv
 â”ƒ â”£ ğŸ“‚MI06A_Residuals_generate
 â”ƒ â”ƒ â”— ğŸ“œRESIDUALS_instances_Age_test.csv
 â”ƒ â”£ ğŸ“‚MI06B_Residuals_correlations
 â”ƒ â”ƒ â”£ ğŸ“œResidualsCorrelations_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œResidualsCorrelations_samplesizes_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œResidualsCorrelations_sd_instances_Age_test.csv
 â”ƒ â”ƒ â”— ğŸ“œResidualsCorrelations_str_instances_Age_test.csv
 â”ƒ â”£ ğŸ“‚MI07A_Select_best
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_bestmodels_alphabetical_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPERFORMANCES_bestmodels_ranked_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œPREDICTIONS_bestmodels_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œRESIDUALS_bestmodels_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œResidualsCorrelations_bestmodels_instances_Age_test.csv
 â”ƒ â”ƒ â”£ ğŸ“œResidualsCorrelations_bestmodels_sd_instances_Age_test.csv
 â”ƒ â”ƒ â”— ğŸ“œResidualsCorrelations_bestmodels_str_instances_Age_test.csv
 â”ƒ â”£ ğŸ“œfake_PA_visit_date.csv
 â”ƒ â”£ ğŸ“œfake_all_eids.csv
 â”ƒ â”£ ğŸ“œfake_short_ukb41230.csv
 â”ƒ â”— ğŸ“œmissing_samples.csv
 â”£ ğŸ“‚scripts
 â”ƒ â”£ ğŸ“œMI01A_Preprocessing_main.py
 â”ƒ â”£ ğŸ“œMI01B_Preprocessing_imagesIDs.py
 â”ƒ â”£ ğŸ“œMI01C_Preprocessing_folds.py
 â”ƒ â”£ ğŸ“œMI02_Training.py
 â”ƒ â”£ ğŸ“œMI03A_Predictions_generate.py
 â”ƒ â”£ ğŸ“œMI03B_Predictions_concatenate.py
 â”ƒ â”£ ğŸ“œMI03C_Predictions_merge.py
 â”ƒ â”£ ğŸ“œMI03D_Predictions_eids.py
 â”ƒ â”£ ğŸ“œMI04A05B_Performances_generate.py
 â”ƒ â”£ ğŸ“œMI04B05C_Performances_merge.py
 â”ƒ â”£ ğŸ“œMI04C_Performances_tuning.py
 â”ƒ â”£ ğŸ“œMI05A_Ensembles_predictions.py
 â”ƒ â”£ ğŸ“œMI06A_Residuals_generate.py
 â”ƒ â”£ ğŸ“œMI06B_Residuals_correlations.py
 â”ƒ â”£ ğŸ“œMI07A_Select_best.py
 â”ƒ â”— ğŸ“œMI_Classes.py
 â”£ ğŸ“œ.gitignore
 â”£ ğŸ“œREADME.md
 â”£ ğŸ“œREADME_installation.md
 â”£ ğŸ“œabdomen_tutorial.sh
 â”— ğŸ“œrequirements.txt